lib/src/main/kotlin/trees/abstractTree.kt

```diff
@@ -0,0 +1,157 @@
+                    changeChild(nodeToDelete, parentNode, nodeToDelete.rightChild)
+                    return nodeToDelete.rightChild
+                }
+                else {
+                    changeChild(nodeToDelete, parentNode, nodeToDelete.leftChild)
+                    return nodeToDelete.leftChild
+                }
+            }
+
+            /* 2 children case */
+            else -> {
+                val replacementNode = findMinNodeInRight(nodeToDelete.rightChild)
+                    ?: throw IllegalArgumentException ("Node with 2 children must have a right child")
+                changeChild(replacementNode, findParent(replacementNode), replacementNode.rightChild)
+                replacementNode.leftChild = nodeToDelete.leftChild
+                replacementNode.rightChild = nodeToDelete.rightChild
+                changeChild(nodeToDelete, parentNode, replacementNode)
+                return replacementNode
+            }
+        }
+    }
+
+    protected fun findParent(node: someNode): someNode? {
+        var curNode = root
+        while (curNode != null) {
+            if (curNode.key == node.key) return null
+            if (curNode.leftChild?.key == node.key || curNode.rightChild?.key == node.key) return curNode
+            curNode = when {
+                curNode.key < node.key -> curNode.rightChild
+                else -> curNode.leftChild
+            }
+        }
+        return null
+    }
+
+    /** rebinds "children" links from one node to another
+     *
+     *        parentNode     -->      parentNode
+     *       /                       /
+     *    node              replacementNode
+     */
+    protected fun changeChild(node: someNode, parentNode: someNode?, replacementNode: someNode?) {
+        when (parentNode) {
+            null -> if (root == node) root = replacementNode
+            else -> {
+                if (parentNode.rightChild == node) parentNode.rightChild = replacementNode
+                else if (parentNode.leftChild == node) parentNode.leftChild = replacementNode
+            }
+        }
+    }
+
+    protected fun findMinNodeInRight(subtree: someNode?): someNode? {
+        var minNode = subtree
+        while (true) {
+            minNode = minNode?.leftChild ?: break
+        }
+        if (minNode != null) return minNode
+        return null
+    }
+
+    fun find(key: K): V? {
+        return findNodeByKey(key)?.value
+    }
+
+    protected fun findNodeByKey(key: K): someNode? {
+        var curNode: someNode? = root ?: return null
+        while (curNode != null) {
+            curNode = when {
+                (curNode.key == key) -> return curNode
+                (curNode.key > key) -> curNode.leftChild
+                else -> curNode.rightChild
+            }
+        }
+        return null
+    }
+
+    protected fun traverse(curNode: someNode?, listOfNodes: MutableList<someNode>) {
+        if(curNode != null) {
```

Comment:
[Пробел](https://kotlinlang.org/docs/coding-conventions.html#control-flow-statements).
```suggestion
        if (curNode != null) {
```